### PR TITLE
[src] Shutdown CUDA online pipeline deterministically

### PR DESCRIPTION
**Pull request by kkm000**
Also reformat string literals that ended up split into too many
pieces over too many lines, and add commentary on the importance
of data member ordering within the class.

### CODE REVIEW

**Comment by kkm000** [comment]
@galv or @hugovbraun, maybe you want to eyeball this one, too.

**Comment by jtrmal** [comment]
LGTM

**Comment by jtrmal** [comment]
Well, actually, I find the PR name confusing, esp. cs the changes are in the object construction :)

File: src/cudadecoder/batched-threaded-nnet3-cuda-online-pipeline.h [hunk 1]
```diff
@@ -66,24 +66,22 @@ struct BatchedThreadedNnet3CudaOnlinePipelineConfig {
...
-                 "The maximum execution batch size. "
-                 "Larger = Better throughput slower latency.");
-    po->Register("num-channels", &num_channels,
-                 "The number of parallel audio channels. This is the maximum "
-                 "number of parallel audio channels supported by the pipeline"
-                 ". This should be larger "
-                 "than max_batch_size.");
+                 "The maximum execution batch size."
+                 " Larger = better throughput, but slower latency.");
+    po->Register("num-channqels", &num_channels,
```
Comment (danpovey) [inline src/cudadecoder/batched-threaded-nnet3-cuda-online-pipeline.h:71]: typo

**Comment by kkm000** [comment]
@jtrmal, it's very straightforward: constructors of members are called in the order they appear in the class, top to bottom, and the destructors are in the reverse order. So I just moved the CudaFst variable, a "passive" object that just frees device memory when destructed, above "active" objects which (potentially) wait for threads to complete in their destructors and (may) need the long-living device representation of the CudaFst during destruction. I don't think the relative order of the other two matters; at the least, they did not affect the enigmatic three-is-a-charm issue #4556. I just think the more deterministic device-host operations are, the better. So this does not really fix anything, just cuts off some branches when reasoning about it (the other PR does the same in a different area).

I'm just dumping out my changes to keep local modifications to the minimum, whether they're helpful or not. The more deterministic the shutdown of the pipeline + the batcher is, the better. This scenario is not covered in the corresponding binary.

File: src/cudadecoder/batched-threaded-nnet3-cuda-online-pipeline.h [hunk 2]
```diff
@@ -415,22 +412,36 @@ class BatchedThreadedNnet3CudaOnlinePipeline {
   // Only used if feature extraction is run on the CPU
   std::vector<std::unique_ptr<OnlineNnet2FeaturePipeline>> feature_pipelines_;
 
-  // HCLG graph : CudaFst object is a host object, but contains
-  // data stored in
-  // GPU memory
+  // Ordering of the cuda_fst_ w.r.t. thread_pool_ and the decoder is important:
+  // order of destruction is bottom-up, opposite to the order of construction.
+  // We want the FST object, which is entirely passive and only frees device
+  // FST representation when destroyed, to survove both the thread pool and the
```
Comment (danpovey) [inline src/cudadecoder/batched-threaded-nnet3-cuda-online-pipeline.h:418]: survive

**Review by danpovey — COMMENTED** [review]
noticed couple typos.   one is channqels

**Comment by kkm000** [comment]
_(Grumbles.)_ That's a VSCode spellchecker for ya. The only one working I could only find. I suspect it used either heuristics or, I am starting to suspect and fear, an ML model. Misses _survove_ and thinks _chanquels_, _recostruct_ and _otther_ are words (I'm a grumpy otter, and I'm unhappy with that!), but yells on some common medium-frequency words. VSCode gives plugins access to clang-assisted pre-parsed syntax (while Emacs essentially uses a beefed-up regex engine), so it can certainly tell a comment or a string from an identifier.

Emacs + flyspell-prog-mode (with ispell backend) is da best! I do normally spell-check code with it after VSCode. It does not think Doxygen comments are comments and must be spell-checked, tho...

File: src/cudadecoder/batched-threaded-nnet3-cuda-online-pipeline.h (hunk 1 above)
Comment (kkm000) [reply src/cudadecoder/batched-threaded-nnet3-cuda-online-pipeline.h]: Done.

File: src/cudadecoder/batched-threaded-nnet3-cuda-online-pipeline.h (hunk 2 above)
Comment (kkm000) [reply src/cudadecoder/batched-threaded-nnet3-cuda-online-pipeline.h]: Done + 2 more.

**Review by galv — APPROVED** [review]
I remember looking into the what-order-are-member-destructors-run rules as well, back in #4517 (since the relevant cuda error appeared in a destructor, but I wasn't sure which one).

This looks fine to me. Please merge. It is probably better to merge this as is. If I find anything unusual in my internal work, I will speak up.

**Comment by kkm000** [comment]
Thanks @galv! Please also have a look at #4569, it's even simpler to review, very regular changes. Mostly CU_SAFE macros everywhere, 4 asserts in a row at the point where the memory is corrupted in #4556, and deterministic shutdown of device streams and H2H pumps. It fixes nothing, only removes a possible synchronization issues from the scope of the issue.

**Pull request merged**